### PR TITLE
Allow weights trained with CuDNNGRU to be used without GPU

### PR DESCRIPTION
**Pull request by bzamecnik**
TL;DR: This PR extends the plain Keras `GRU` layer with a CuDNN-compatible mode, so that we can reuse weights trained on GPU with `CuDNNGRU` for inference on CPU.

Since #8860 (#8307) we're able to train a LSTM layer quickly on GPU with CuDNN and load the weights to plain LSTM implementation which runs also on a CPU. Since GRU in CuDNN is based on a slightly different convention we can't directly load its weights in a similar way. A possible solution is to provide implementation of GRU layer in plain Keras (similar to existing one) but using the same convention as in CuDNN, so that we can reuse the weights.

As I [described in detail](https://github.com/keras-team/keras/pull/8307#issuecomment-354652830) there are two conventions of GRU:

- most common (Keras, TensorFlow, etc.): hidden input first multiplied by reset gate, then projected by matrix 
- original paper (CuDNN): hidden input first projected, then multiplied by reset gate

In addition CuDNN uses two sets of biases, while Keras just one set.

Note that default `recurrent_activation` (for gates) in Keras is `hard_sigmoid`, while CuDNN uses `sigmoid`, so we should specify that option explicitly.

In this PR we're adding a mode to the `GRU` layer (resp. its `GRUCell`) to use the latter CuDNN-compatible convention and code to convert weights the analogously to the LSTM case (#8307).

Making a separate class for GRU layer or cell allowed easier development/debugging, but results in much code duplication and worse usage, so in the end I just added one more boolean parameter to `GRUCell` to switch between code paths: `variant` which can take values 'reset_before' (reset before multiplication, default), 'reset_after' (reset after multiplication - CuDNN-compatible).

`GRUCell` has already two `implementation`s, one simple, other with fused matrices where possible. Both have been adapted to both conventions. So we have four modes in total.

A basic test for converting weights from CuDNNGRU to GRU is available. Besides that the code has been manually tested on the IMDB RNN example to obtain same results up to numerical precision.

Documentation has been extended to cover this case and reference also the original paper.

UPDATE: I've made a notebook with example of using the code on IMDB example and measuring precision of results and speedup: https://gist.github.com/bzamecnik/bd3786a074f8cb891bc2a397343070f1.

Request for comments:
- Mixing two variants and two implementations is not perfectly readable, but different classes result in too much duplication. At least I tried to deduplicate common code between the two variants. Can you imagine a better way how to structure the code?
- Do you think of any better names for the flag and its values?

Thanks.

UPDATE - results of discussion:

- new `GRU` argument is `reset_after` boolean, default `False`, indicates using the original CuDNN-compatible convention
- shape of stored weights is the same (compared to CuDNNGRU) for kernels, but different for biases
  - ordering of individual weights is different and conversion is possible
  - biases (for input and recurrent kernels) are of shape `(2, 3 * units)`, while in CuDNNGRU, they're flat
    - this is the least bad way to allow distinguishing between weights from `GRU(reset_after=True)` and `CuDNNGRU` since we do not store layer metadata beyond name (such as type or arguments)
- it's possible to convert weights between `GRU(reset_after=True)` and `CuDNNGRU` and also `CuDNNLSTM` and `LSTM` in both directions!

### CODE REVIEW

File: keras/layers/recurrent.py [hunk 1]
```diff
@@ -1197,6 +1200,7 @@ def __init__(self, units,
                  dropout=0.,
                  recurrent_dropout=0.,
                  implementation=1,
+                 variant='reset_before',
```
Comment (fchollet) [inline keras/layers/recurrent.py]: Seems like `reset_after=False` would be more user friendly (and consistent with the implementation). But "reset_after" could also be changed to something more explicit. Thoughts?

Comment (bzamecnik) [reply keras/layers/recurrent.py]: Yeah, that sounds good. I'm not aware of any other conventions than those two, so a boolean might be sufficient.

File: keras/engine/topology.py [hunk 2]
```diff
@@ -3063,6 +3063,29 @@ def preprocess_weights_for_loading(layer, weights,
...
+    # convert the weights of CuDNNGRU so that they could be loaded into
+    if layer.__class__.__name__ == 'GRU' and len(weights) == 3:
+        # determine if we're loading a CuDNNGRU layer from the number of bias weights:
+        # CuDNNGRU has (units * 6) weights; while GRU has (units * 3)
+        # if there's no bias weight in the file, skip this conversion
+        # TODO: refactor with the LSTM conversion code above
+        units = weights[1].shape[0]
+        bias = weights[2]
+        n_gates = 3
+        if len(bias) == units * n_gates * 2:
```
Comment (myutwo150) [inline keras/engine/topology.py]: What if we're loading a `GRU` layer with `reset_after=True`? It will have 6 biases also. I think we shouldn't transpose the kernels in this case.

File: keras/layers/recurrent.py [hunk 3]
```diff
@@ -1468,6 +1534,7 @@ def __init__(self, units,
                  go_backwards=False,
                  stateful=False,
                  unroll=False,
+                 reset_after=True,
```
Comment (myutwo150) [inline keras/layers/recurrent.py]: This default value will break loading old model files.

Comment (mark86092) [reply keras/layers/recurrent.py]: The default value of GRUCell is `reset_after=False`

Comment (fchollet) [reply keras/layers/recurrent.py]: Both should be the same, `False`. This would be a breaking change (besides the fact that both values must be consistent).

Comment (bzamecnik) [reply keras/layers/recurrent.py]: Sorry a silly typo. I did the changes too quickly. I've been flying the whole day and try to fix it later today.

File: keras/engine/topology.py (hunk 2 above)
Comment (bzamecnik) [reply keras/engine/topology.py]: It might be the cause of failed test:
```
tests/keras/engine/test_topology.py::test_preprocess_weights_for_loading[layer0] FAILED
```

File: keras/layers/recurrent.py (hunk 3 above)
Comment (bzamecnik) [reply keras/layers/recurrent.py]: It caused failed test:
```
tests/keras/engine/test_topology.py::test_preprocess_weights_for_loading[layer0] FAILED
```

File: keras/engine/topology.py (hunk 2 above)
Comment (bzamecnik) [reply keras/engine/topology.py]: This is a really good point. There's a problem that the loading code has no clue how to distinguish weights from `CuDNNGRU` and `GRU(reset_after=True)` since they have the same shape and there are not metadata in the HDF file.

Comment (bzamecnik) [reply keras/engine/topology.py]: I can see some ways how to make those cases distinguishable:

- save some additional metadata along the weights, such as the class name or parameters
  - we cannot expect that from existing weight files
  - if we want consistency, we'd have to do it for all layers - quite a big change
- change the shape saved from `GRU(reset_after=True)` so that it's a bit different from `CuDNNGRU` (an ugly hack)
- try to infer the from the weight contents - IMHO impossible

Comment (myutwo150) [reply keras/engine/topology.py]: It seems like the `weights` argument for `preprocess_weights_for_loading` is not really a list of numpy arrays. It's a list of `HDF5 dataset`. When I call `print(weights)` on my machine, I saw something like:
```
[<HDF5 dataset "kernel:0": shape (100, 96), type "<f4">,
 <HDF5 dataset "recurrent_kernel:0": shape (32, 96), type "<f4">,
 <HDF5 dataset "bias:0": shape (192,), type "<f4">]
```

Hence there would be some metadata attached to the weights. Specifically, with `print([x.name for x in weights])`, the output is:
```
['/model_weights/cu_dnngru_1/cu_dnngru_1/kernel:0',
 '/model_weights/cu_dnngru_1/cu_dnngru_1/recurrent_kernel:0',
 '/model_weights/cu_dnngru_1/cu_dnngru_1/bias:0']
```

However, I'm really not sure if this is a behavior that we could rely on (i.e., is it a consistent behavior for different versions of h5py, different versions of Keras, python, OS, ...). Also, it might not pass some existing tests since the tests are written under the assumption that the input is a list of arrays.

Comment (bzamecnik) [reply keras/engine/topology.py]: What's also missing are two other cases:

- `GRU(reset_after=True)` -> `CuDNNGRU` - needs reverse conversion
- `GRU(reset_after=True)`/`CuDNNGRU` <-> `GRU(reset_after=False)` - incompatible, show fail fast with a meaningful message

It means that we could also possibly support conversion from LSTM to CuDNNLSTM (in a separate PR).

Comment (myutwo150) [reply keras/engine/topology.py]: I agree. This solution sounds good to me.

**Comment by bzamecnik** [comment]
Failed tests fixed. Problem with distinguishing between weights from `GRU(reset_after=True)` and `CuDNNGRU` fixed by changing shape of biases of the GRU layer. Tests have been added and improved. Incompatible conversions fail fast. As a bonus it's possible to load weights from `GRU(reset_after=True)` back to `CuDNNGRU`. I'd like to add analogous conversion for LSTM.

**Comment by bzamecnik** [comment]
Now Travis tests pass, except a [failure on a possibly unrelated out-of-memory error](https://travis-ci.org/keras-team/keras/jobs/337251772) when forking in `tests/keras/utils/data_utils_test.py::test_finite_generator_enqueuer_processes` in python2.7/tensorflow:

```
>       self.pid = os.fork()
E       OSError: [Errno 12] Cannot allocate memory
```

File: keras/engine/topology.py [hunk 4]
```diff
@@ -3042,26 +3042,107 @@ def preprocess_weights_for_loading(layer, weights,
...
-    # convert the weights of CuDNNLSTM so that they could be loaded into LSTM
-    if layer.__class__.__name__ == 'LSTM' and len(weights) == 3:
+    weights = _convert_rnn_weights(layer, weights)
+
+    return weights
+
+
+def _convert_rnn_weights(layer, weights):
+    """
+    Converts weights for RNN layers between native and CuDNN format.
```
[the remark this answers is not in the document]
Comment (bzamecnik) [reply keras/engine/topology.py]: Do you mean making a longer function name (like `_convert_rnn_weights_native_cudnn()`) instead of a comment?

[the remark this answers is not in the document]
Comment (myutwo150) [reply keras/engine/topology.py]: Ah sorry, I mean:
```
"""Converts weights for RNN layers between native and CuDNN format.
"""
```
I think I've seen this suggestion many times in PR comments.

[the remark this answers is not in the document]
Comment (bzamecnik) [reply keras/engine/topology.py]: Aah. Understand. I haven't seen this more inline format of `"""` docs. OK.

File: tests/keras/layers/cudnn_recurrent_test.py [hunk 5]
```diff
@@ -463,5 +474,64 @@ def test_cudnnrnn_bidirectional():
     model.fit(x, y, epochs=1, batch_size=1)
 
 
+@keras_test
+@pytest.mark.parametrize('layer_class,layer_args', [
+    (keras.layers.GRU, {'units': 2, 'input_shape': [3, 5]}),
+    (keras.layers.GRU, {'units': 2, 'input_shape': [3, 5], 'reset_after': True}),
+    (keras.layers.LSTM, {'units': 2, 'input_shape': [3, 5]}),
```
[the remark this answers is not in the document]
Comment (bzamecnik) [reply tests/keras/layers/cudnn_recurrent_test.py]: It would be good to keep the non-cudnn stuff back in `test_topology.py` and call the test method with `skipif` and different args for cudnn layers, however I'm not sure if such cross-test import would work since `tests/` do not have module structure. If we only want to allow running plain GRU/LSTM variants on Travis, we can just split it into two functions within the same file, possibly just the original `test_topology.py`, even though some classes would be cudnn-related.

[the remark this answers is not in the document]
Comment (bzamecnik) [reply tests/keras/layers/cudnn_recurrent_test.py]: Fixed.

For some strange reason`test_preprocess_weights_for_loading_rnn_should_be_idempotent()` with GRU `{'reset_after': True}` fails when called without test parallelization `py.test -n 0 tests/keras/engine/test_topology.py::test_preprocess_weights_for_loading_rnn_should_be_idempotent` on `reset_after` being propagated to RNN kwargs. Without `-n 0` it behaves as expected, ie. `reset_after` is consumed as a named arg and not in kwargs. WTF? :)

**Comment by bzamecnik** [comment]
@myutwo150 All review comments fixed. Thanks.

File: keras/layers/recurrent.py [hunk 6]
```diff
@@ -1238,31 +1243,56 @@ def build(self, input_shape):
...
+                bias_shape = (2, 3 * self.units)
+            self.bias = self.add_weight(shape=bias_shape,
                                         name='bias',
                                         initializer=self.bias_initializer,
                                         regularizer=self.bias_regularizer,
                                         constraint=self.bias_constraint)
+            if not self.reset_after:
+                self.input_bias, self.recurrent_bias = self.bias, None
+            else:
+                self.input_bias, self.recurrent_bias = self.bias[0], self.bias[1]
```
Comment (myutwo150) [inline keras/layers/recurrent.py]: The CNTK errors on Travis CI seem to come from this line. `self.bias[0]` results in a `(1, 3 * self.units)` 2-D tensor in CNTK.

Comment (bzamecnik) [reply keras/layers/recurrent.py]: Aah. Thanks!

Comment (bzamecnik) [reply keras/layers/recurrent.py]: So it means slicing in CNTK is not numpy compatible, right?

Comment (myutwo150) [reply keras/layers/recurrent.py]: Yes, the behavior is different from numpy.

Comment (bzamecnik) [reply keras/layers/recurrent.py]: I filed a bug report https://github.com/Microsoft/CNTK/issues/2938. In the meanwhile we have to find a way how to slice it to 1D. API docs is not helpful.

Comment (bzamecnik) [reply keras/layers/recurrent.py]: Ok, solution seems to be:`K.reshape(self.bias[0], (-1,))` to ensure 1D shape.

**Comment by bzamecnik** [comment]
It seems `K.flatten` fixed the problem with CNTK. Now only TensorFlow/Python 3.6 failed while downloading TensorFlow package: `ConnectionResetError: [Errno 104] Connection reset by peer`.

File: keras/engine/topology.py [hunk 7]
```diff
@@ -3042,26 +3042,104 @@ def preprocess_weights_for_loading(layer, weights,
...
 
-    # convert the weights of CuDNNLSTM so that they could be loaded into LSTM
-    if layer.__class__.__name__ == 'LSTM' and len(weights) == 3:
+    weights = _convert_rnn_weights(layer, weights)
+
+    return weights
+
+
+def _convert_rnn_weights(layer, weights):
+    """Converts weights for RNN layers between native and CuDNN format.
```
Comment (fchollet) [inline keras/engine/topology.py]: Please format the docstring like the others, with an `# Arguments` and `# Returns` section

File: keras/engine/topology.py [hunk 8]
```diff
@@ -3042,26 +3042,104 @@ def preprocess_weights_for_loading(layer, weights,
...
+    """
+
+    def transform_kernels(kernels, func, n_gates):
```
Comment (fchollet) [inline keras/engine/topology.py]: Likewise, please format docstring

File: keras/engine/topology.py [hunk 9]
```diff
@@ -3042,26 +3042,104 @@ def preprocess_weights_for_loading(layer, weights,
...
+        """Transforms kernel for each gate separately using given function.
+        """
+        return np.hstack([func(k) for k in np.hsplit(kernels, n_gates)])
+
+    def transpose_input(from_cudnn):
+        """Transforms input kernels from/to CuDNN format.
```
Comment (fchollet) [inline keras/engine/topology.py]: Likewise, please format docstring

**Comment by abhiboost** [comment]
is this going to be in 2.1.5?

**Comment by abhiboost** [comment]
I get the following error when I try to load the weights of CuDNNGRU to GRU(reset_after=True):

```
  File "/usr/local/lib/python3.5/dist-packages/keras/engine/topology.py", line 2667, in load_weights
    f, self.layers, reshape=reshape)
  File "/usr/local/lib/python3.5/dist-packages/keras/engine/topology.py", line 3380, in load_weights_from_hdf5_group
    reshape=reshape)
  File "/usr/local/lib/python3.5/dist-packages/keras/engine/topology.py", line 3004, in preprocess_weights_for_loading
    original_backend)
  File "/usr/local/lib/python3.5/dist-packages/keras/engine/topology.py", line 3156, in preprocess_weights_for_loading
    weights = _convert_rnn_weights(layer, weights)
  File "/usr/local/lib/python3.5/dist-packages/keras/engine/topology.py", line 3295, in _convert_rnn_weights
    weights = convert_weights(weights, from_cudnn=True)
  File "/usr/local/lib/python3.5/dist-packages/keras/engine/topology.py", line 3270, in convert_weights
    biases = weights[2].reshape((2, -1) if from_cudnn else -1)
AttributeError: 'Dataset' object has no attribute 'reshape'

```

**Comment by matfiore** [comment]
@bzamecnik @fchollet  I encountered the same problem as @abhiboost and the fix  proposed in #9662 by @myutwo150 works for me!

**Comment by bzamecnik** [comment]
Also check #10080 (CuDNN RNN layers nested in Model are not converted when loading) for nested Models.

**Comment by azraelkuan** [comment]
HI, i am using the keras layers in the tensorflow, i found that when loading weight from CuDNNGRU to GRU using model.load_weight(), it is ok.
but if i use the tf.train.Saver to restore, the bias will be restored as [num_units*6, ] will not be first reshape to [2, 3 * num_units], this is a bug?
